csrc/parallel_dimension_map.cpp

```diff
@@ -17,169 +18,59 @@
+        dim_map_[ptype] = concrete_id->extent();
+      } else {
+        dim_map_.at(ptype) = SimplifyingIrBuilder::maxExpr(
+            dim_map_.at(ptype), concrete_id->extent());
+      }
+    }
   }
-}
-
-void ParallelDimensionMap::populateDimensionMapWithSingleCASet(
-    ParallelType pt,
-    const std::unordered_set<IterDomain*>& dom_set) {
-  TORCH_INTERNAL_ASSERT(dom_set.size() == 1);
 
-  // pt is used by only one concrete domain
-  auto id = *dom_set.begin();
-  auto it = constant_extent_map_.find(id);
-
-  if (it != constant_extent_map_.end()) {
-    TORCH_INTERNAL_ASSERT(
-        it->second.size() == 1,
-        "Only one value found mapped to parallel type ",
-        stringifyThread(pt),
-        " yet its bound to multiple extents.");
-    dim_map_.insert({pt, IrBuilder::create<Int>(*(it->second.begin()))});
-    exact_types_.insert(pt);
-  } else {
-    // Prefer to use blockDim/gridDim if not constant
-    dim_map_.insert({pt, NamedScalar::getParallelDim(pt)});
-    exact_types_.insert(pt);
+  // Simplify dim_map_
+  for (auto& [k, v] : dim_map_) {
+    v = simplifyExpr(v);
   }
-}
-
-void ParallelDimensionMap::populateDimensionMapWithMultipleCASet(
-    ParallelType pt,
-    const std::unordered_set<IterDomain*>& dom_set) {
-  TORCH_INTERNAL_ASSERT(dom_set.size() > 1);
-
-  bool all_equal = true;
-  // Use nullptr to signal it's not initialied yet
-  Val* known_dimension = nullptr;
-  // Use -1 to signal it's not initialied yet
-  int64_t known_const = -1;
 
-  // Check all of concrete domains to see if they match all together.
-  for (auto concrete_id : dom_set) {
-    if (concrete_id->isBroadcast()) {
-      // Broadcasted concrete id's don't specify anything about shape
-      continue;
-    }
-    // If this concrete domain has a constant extent, check if it
-    // matches with the known constant extent.
-    auto it = constant_extent_map_.find(concrete_id);
-    if (it != constant_extent_map_.end()) {
-      const auto& const_extent_set = it->second;
-      // If multiple constants are detected, it's not exact.
-      if (const_extent_set.size() > 1) {
-        all_equal = false;
-        break;
+  // Compute exact_types_
+  for (auto tv : ir_utils::filterByType<TensorView>(all_vals)) {
+    for (auto id : tv->domain()->domain()) {
+      auto ptype = id->getParallelType();
+      if (exact_types_.count(ptype) == 0) {
+        continue;
       }
-      auto this_const = *(const_extent_set.begin());
-      // known_const is initialized to -1
-      if (known_const == -1) {
-        known_const = this_const;
-      } else if (known_const == this_const) {
-        // Matched with previously known const. The extent of this
-        // domain must be equal to that's previously known.
+      auto concrete_id = GpuLower::current()->caMap()->getConcreteMappedID(
+          id, IdMappingMode::EXACT);
+      if (concrete_id->isBroadcast()) {
+        // Broadcasted concrete id's don't specify anything about shape
```

Comment:
I'm not sure I agree with this. If a concrete ID is a broadcast ID and parallelized, we would need to have a predicate like `threadIdx.x < 1` if `threadIdx.x` is not exact. Is it safe to ignore concrete broadcast IDs?

Reply:
My understanding is this won't cause any IMA or wrong result, but it might cause redundant computation (some thread computes something and then abandon it). I am not strong on ignoring concrete broadcast, but isn't this behavior closest to the original `populateDimensionMapWithMultipleCASet`?

Reply:
Ha, you're right. We already had the same comment!